desc/compute/_profiles.py

```diff
@@ -1474,3 +1474,32 @@ def _current_r(params, transforms, profiles, data, **kwargs):
 def _current_rr(params, transforms, profiles, data, **kwargs):
     data["current_rr"] = 2 * jnp.pi / mu_0 * data["I_rr"]
     return data
+
+
+@register_compute_fun(
+    name="shear",
+    label="\\rho \\partial_{\\rho} \\iota / \\iota",
+    units="~",
+    units_long="None",
+    description="Normalized derivative of the rotational transform",
+    dim=1,
+    params=[],
+    transforms={},
+    profiles=[],
+    coordinates="r",
+    data=["0", "iota", "iota_r", "iota_rr", "rho"],
+)
+def _shear(params, transforms, profiles, data, **kwargs):
+    eps = 1e2 * jnp.finfo(data["iota"].dtype).eps
+    data["shear"] = cond(
+        jnp.all(jnp.abs(data["iota"])) < eps,
+        lambda _: data["0"],  # if iota profile is all 0, set shear to 0
+        lambda _: jnp.where(  # apply l'Hopital rule where necessary
+            (jnp.abs(data["iota"]) < eps)
+            & ((jnp.abs(data["rho"]) < eps) | (jnp.abs(data["iota_r"]) < eps)),
+            1 + data["rho"] * data["iota_rr"] / data["iota_r"],  # limit
+            data["rho"] * data["iota_r"] / data["iota"],  # shear
+        ),
+        None,
+    )
+    return data
```

Comment:
Typically $\partial_{\rho} \iota$ is 0 at axis because it should be able to be expanded as an even power series near axis so that it is symmetric under the transformation $\rho \to -\rho$. So in that case in order to compute this limit we need an additional application of lhipitals rule which yields $\lim_{\rho \to 0} \rho \frac{\iota}{\partial_{\rho} \iota} = 2 + \rho \frac{\partial_{\rho \rho \rho} \iota}{\partial_{\rho \rho} \iota}$.

We don't have the third derivative of $\iota$ ready to compute, so I think we should only handle the case where $\iota = 0$ everywhere and add `shear` to the not implemented limits list: https://github.com/PlasmaControl/DESC/blob/c0b44414883d202fb30c94e78fbbd5e7d29b59d7/tests/test_axis_limits.py#L53

Reply:
Also since the limit is undefined in the case iota = 0 everywhere, and I think we are just setting shear to 0 in this case to avoid `nan` appearing in the objective, I feel like it would be better to do this inside the objective function rather than the compute function.

If you choose to continue handling this inside the compute function, can you add a comment that says we set it to zero for numerical reasons and it's not the actual limit

Reply:
Thanks for typing this out. I still think my comments above are valid.

> One exception is for Reverse Field Configurations (RFCs), where the shear reverses sign and crosses through $\iota=0$. But even then, this typically happens towards the last closed flux surface and far from the magnetic axis. In this case I think the shear is undefined because $\iota=0$ but $\iota'\neq0$, and our compute function will return `NaN`. 

Yes I agree.

> The only other exception is a vacuum tokamak, which has $\iota(\rho)=0~\forall\rho$ and we already cover that case. Otherwise any amount of torsion, 3D shaping, or toroidal current should give a finite rotational transform on the magnetic axis. 

In this case, the shear: $\rho \frac{\iota'}{\iota}$ is undefined everywhere (including axis) because there is no information about the speed the function approaches any point. At any point, shear is equivalent to the undefined arithmetic operation $0 / 0$ times $\rho$.  Currently we arbitrarily say this limit is 0. I'm guessing this is for numerical purposes of  avoiding  `nan` in the objective?

> I think the actual case we implemented the limit for where $\iota=\iota'=0$ is very rare. 

Yes it seems rare to me too. We failed to implement the limit in this case. In this case the code returns $1 + \rho \frac{\iota''}{\iota'}$. This will evaluate as `nan` because $\iota'$ is 0. The correct limit is

- 1 if $\rho = 0$ and $\iota'' \neq 0$.
- If $\rho = 0$ and $\iota'' = 0$, we need a second application of lhopitals rule which gives $2 + \rho \frac{\partial_{\rho \rho \rho} \iota}{\partial_{\rho \rho} \iota}$. This is also indeterminate so we need a third application $3 + \rho \frac{\partial_{\rho \rho \rho \rho} \iota}{\partial_{\rho \rho \rho} \iota}$. If this last term is not indeterminate we are done. Otherwise we continue applying lhopitals.
- If $\rho \neq 0$ and $\iota'' = 0$, then the same thing as above happens if $\partial_{\rho \rho \rho} \iota = 0$ too. Otherwise the limit is probably not finite.
- If $\rho \neq 0$ and $\iota'' \neq 0$, then the limit is probably not finite. 


This seems pedantic enough that I suggest only implementing what you need for the objective in the vacuum case and adding a comment that it's not the true limit. If I've made a mistake, sorry for using your time.

Reply:
This is an important conversation, not a waste of time! I misunderstood your earlier comment, and I now see your point that the limit as implemented was useless. 

Yes, we "arbitrarily" set `shear = 0` when $\iota(\rho)=0~\forall\rho$ to avoid `NaN`, because starting from a vacuum tokamak is such a common use case. It is a physics-informed choice, even if it is not mathematically rigorous. 

I've decided to remove the handling of the limits for other cases where $\iota=0$, but added comments in the code to explain. As we have discussed, the shear *should* be undefined for most of these instances such as in an RFC. And as you pointed out, if $\iota=\iota'=0$ we would need higher order derivates, which I do not think is worth implementing for such a rare edge case.